usbc: driver for parade ps8xxx series TCPCs

barnas-michal opened the pull request.

barnas-michal committed:
usbc: change type for functions that may return error code

TCPC API functions should be able to inform the caller if the function
is supported and successfully executed. Bool values doesn't allow this
so it is needed to change the type to int.
For is_rx_pending_msg function the return code should conforms to
existing error codes, so in case of functino being not supported,
the -ENOSYS should be returned. In case of successful execution,
if there is no pending message, the -ENODATA should be returned and
in case of message pending, the value of 0.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

sambhurst commented:
Hello @semihalf-barnas-michal, could you provide instruction on how you testing this PR?

sambhurst commented:
The following three CLs should be in a separate PR. They are not USB-C specific and should be reviewed by a wider audience.
[i2c: use device instead of name for i2c dump messages](https://github.com/zephyrproject-rtos/zephyr/pull/60224/commits/98c16f4807b4dffaea3a7628a21e6404e8e7eb07)
[i2c: improve the message dumps messages](https://github.com/zephyrproject-rtos/zephyr/pull/60224/commits/9bbff828fd0350345c701de6d33556d49779ea9f)
[i2c: add filtering of i2c dumped messages ](https://github.com/zephyrproject-rtos/zephyr/pull/60224/commits/2eeb4d0a98a741d7fc39d7af0b79c8bc8fdd7626)

cybekRT commented:
> could you provide instruction on how you testing this PR?

You need to get brya and flash with firmware from this CL: https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/4669489
Then, after booting the zephyr, you should execute the shell command (in EC console) "startpd", this command starts the usb-c stack, after this, the right usb-c port should be detecting the charger and gets the second PDO, probably 9 or 12V.

> The following three CLs should be in a separate PR

They are separeted here: https://github.com/zephyrproject-rtos/zephyr/pull/60301
This PR was just done before separating, it will be rebased later.

sambhurst commented on subsys/usb/usb_c/usbc_pe_common.c:
```diff
@@ -345,6 +345,8 @@ void pe_message_received(const struct device *dev)
 	struct policy_engine *pe = data->pe;
 
 	atomic_set_bit(pe->flags, PE_FLAGS_MSG_RECEIVED);
+
```
Add a comment about why this is needed. Same with the others.

sambhurst commented on subsys/usb/usb_c/usbc_stack.c:
```diff
@@ -113,7 +115,10 @@ int usbc_suspend(const struct device *dev)
 	data->request.val = PRIV_PORT_REQUEST_SUSPEND;
-	k_fifo_put(&data->request_fifo, &data->request);
+
+	if (k_fifo_is_empty(&data->request_fifo)) {
```
If you are preventing the subsystem from suspending, you should return a value that informs the caller
of the failure to suspend. Maybe return -EBUSY.

semihalf-barnas-michal replied:
It's not exactly that I am preventing the suspend. The request is inside the data object of the usb-c port instance, so only one request can exist in the same time. This function overrides the value of the request, so theoretically the request should still be valid (but it should be verified further to check for any race conditions) and overrides the previous one. Main problem can be if we use the REQUEST_NOP for state changing and other thread wants to suspend the usb-c, it needs more checks, or just check if overwritten request was not a suspend.

But the main problem here is that we have only one instance of the request object, so if the fifo is not empty, that means this object was already added to the fifo, and adding it again causes the deadlock during the iteration of the fifo by the usb-c stack.

semihalf-barnas-michal replied:
As we discussed, this was changed to more specific solution used entirely to prevent one thread sleep to improve the stack response time. Previous solution was a misuse of the existing logic.

sambhurst commented on subsys/usb/usb_c/usbc_stack.c:
```diff
@@ -128,7 +133,12 @@ int usbc_request(const struct device *dev, const enum usbc_policy_request_t req)
 	/* Add public request to fifo */
 	data->request.val = req;
-	k_fifo_put(&data->request_fifo, &data->request);
+	if (k_fifo_is_empty(&data->request_fifo)) {
```
You should return an appropriate error code, informing the caller that the request was not successful. Also, how do you know that the same request is being added?

semihalf-barnas-michal replied:
This case was a misuse of the usbc_request function by my changes. I have added function dedicated to preventing one sleep of USB-C thread causing the one more iteration of the state machines.

sambhurst commented on subsys/usb/usb_c/usbc_tc_snk_states.c:
```diff
@@ -79,7 +79,11 @@ static void sink_power_sub_states(const struct device *dev)
+
 	LOG_INF("Unattached.SNK");
+
+	usbc_request(tc->dev, REQUEST_NOP);
```
Add comment

sambhurst commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -401,6 +401,8 @@ static void prl_tx_set_state(const struct device *dev, const enum usbc_prl_tx_st
 	__ASSERT(state < ARRAY_SIZE(prl_tx_states), "invalid prl_tx_state %d", state);
 	smf_set_state(SMF_CTX(prl_tx), &prl_tx_states[state]);
+
+	usbc_request(dev, REQUEST_NOP);
```
Add comment

sambhurst commented on drivers/usb_c/tcpc/ucpd_stm32.c:
```diff
@@ -1027,10 +1027,9 @@ static int ucpd_transmit_data(const struct device *dev,
  * @brief Tests if a received Power Delivery message is pending
  *
- * @retval true if message is pending, else false
+ * @retval 1 if message is pending, else 0
```
I think this can stay true and false

sambhurst replied:
Sorry, I see why you changed this to 1 and 0. I think there should be a better way to handle these return values and error codes. Because an error code can result, maybe we just stick with convention and return 0 on success and a negative error code on failure for the functions you've modified in this CL. This would mean all the places these functions are used; those return value checks would need updating.

semihalf-barnas-michal replied:
Hm, I think about adding the additional parameter which will be a pointer to bool that could contain the status of pending message. I don't think that using the "there's no pending message" as error code is valid, since it's not an error. However, I see that the _tcpc_receive_data_ is using the concept of returning negative values as errors and in case of success, the zero of other positive number. So it would be according to this concept of returning either negative values, or 0/1 for value got.

sambhurst commented on drivers/usb_c/tcpc/Kconfig.tcpc_tcpci:
```diff
@@ -0,0 +1,36 @@
+	default 2
+	help
+	  Number of I2C transaction tries that will be performed for each request.
+	  Some TCPCs are going into deep sleep mode when no charger is connected and won't respond
```
formatting

semihalf-barnas-michal replied:
What do you mean? If you think about text being too long, the zephyr uses the 100 columns per line instead of 80

sambhurst commented on drivers/usb_c/tcpc/tcpci.c:
```diff
@@ -0,0 +1,368 @@
+
+#define LOG_COMM_ERR_STR "Can't communicate with TCPC %s@%x (%s %x = %04x)"
+
+const struct tcpci_reg_dump_map tcpci_std_regs[TCPCI_STD_REGS_SIZE] = {
```
add comment @brief

sambhurst commented on drivers/usb_c/tcpc/tcpci.c:
```diff
@@ -0,0 +1,368 @@
+	},
+};
+
+int tcpci_read_reg8(const struct i2c_dt_spec *i2c, uint8_t reg, uint8_t *value)
```
Add comment @brief and the same for the other function.

semihalf-barnas-michal replied:
Is there a need to add functions documentation to source files if header or API header contains them? I see that the ucpd driver contains documentation for its implementation functions duplicated from the _usbc_tcpc.h_. And the drivers functions should be compliant to the API functions and their documentation.

sambhurst replied:
I understand, but I don't know if it's convention to have a brief explanation of the function here too. Other reviewers, please comment if it's okay to just document the functions in the header file.

sambhurst commented on drivers/usb_c/tcpc/tcpci_shell.c:
```diff
@@ -0,0 +1,31 @@
+
+#define DUMP_TCPC_REGS(node) ret |= tcpc_dump_std_reg(DEVICE_DT_GET(DT_PROP(node, tcpc)));
+
+static int cmd_tcpci_dump(const struct shell *sh, size_t argc, char **argv)
```
add comment @brief

sambhurst commented on drivers/usb_c/vbus/usbc_vbus_tcpci.c:
```diff
@@ -0,0 +1,109 @@
+#include <zephyr/usb_c/tcpci.h>
+#include <zephyr/shell/shell.h>
+
+struct vbus_tcpci_cfg {
```
Add comment

sambhurst commented on drivers/usb_c/vbus/usbc_vbus_tcpci.c:
```diff
@@ -0,0 +1,109 @@
+	const struct i2c_dt_spec i2c;
+};
+
+struct vbus_tcpci_data {
```
Should this be empty?

semihalf-barnas-michal replied:
I left it at the beginning in case I would need some variables. Will remove it.

sambhurst commented on include/zephyr/drivers/usb_c/tcpci_priv.h:
```diff
@@ -0,0 +1,29 @@
+#include <stdint.h>
+#include <zephyr/drivers/i2c.h>
+
+struct tcpci_reg_dump_map {
```
Add comment

sambhurst commented on include/zephyr/usb_c/tcpci.h:
```diff
@@ -0,0 +1,285 @@
+#ifndef ZEPHYR_INCLUDE_USB_C_TCPCI_H_
+#define ZEPHYR_INCLUDE_USB_C_TCPCI_H_
+
+#define TCPC_REG_VENDOR_ID  0x0
```
These should all be commented

barnas-michal committed:
usbc: add check if tcpc initialization was successful

The TCPC driver initialization can be unsuccessful and it should be
checked and logged if that's the case.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: check if message is pending before trying to receive it

If TCPC supports checking for pending messages, it should be done
instead of forcefully trying to read the message and getting no-data
result. This helps with powereficiency for chips that go to sleep
when there is no message pending. Trying to read the message can
wake up the chip providing higher power usage.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add setter function for sink and source control

Add set_snk_ctrl and set_src_ctrl functions to TCPC API to allow
controlling the sourcing and sinking vbus.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: check if sampling CC lines were successful

When the get_cc function returns error, the CC lines should be
considered as open. It may happen either due to some chip malfunction
or chip going into sleep mode and needing time to reinitialize.
TCPC should only go to sleep if there is no partner attaches, so
in case of this error, the lines should be considered as disconnected.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: check if setting the CC and VCONN was successful

This commit adds check in cc_open entry state and logs if setting
either of the values were unsuccessful.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add support for TCPC control of vbus sourcing

Adds calls to the set_src_ctrl of TCPC driver to enable and disable
the VBUS sourcing. If the TCPC doesn't support these functions,
no errors should be printed.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add function that prevents the USB-C stack from sleep

Add internal function that prevents one sleep of the USB-C stack.
It can allow to have longer sleep times to conserve more power, while
still having possibility to respond in the required time and
set specific registers.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add sleep prevention in places that may require it

Add sleep prevention in crucial USB-C stack functions to make
state transitions faster and to send responses faster.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add definitions for TCPCI registers and fields

Add header file with registers and fields definitions compliant
to the Type-C Port Controller Interface Specification
Revision 2.0, Version 1.3.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add generic TCPCI related functions

Add generic functions that will be common to all TCPCI compliant
drivers like registers reading, writing and updating.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add support for vbus measurement using TCPCI compliant device

Add support for VBUS measuring part of the TCPCI compliant device.
This device should be used as a child-node for the more specific
TCPC driver and referenced by the vbus property in the usb-c
connector node.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add "tcpc" shell command

Add "tcpc" shell command with "dump" subcommand used to dump
all registers of all the TCPCs.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add "vbus" subcommand for "tcpc" shell command

Add "vbus" subcommand that displays VBUS voltages for all USB-C
connectors.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add "chip" subcommand for "tcpc" shell command

Add "chip" subcommand that displays the vendor, product and device
identifiers for all TCPC referenced by USB-C connectors.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add TCPC driver for Parade PS8815 chip

Add support for the PS8815 and used with other PS8xxx family chips.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add event-driven handling of CC lines changes in ps8815 driver

Add bool value that stores the CC change information received in
the interrupt of TCPC alert line. In sink role when in disconnected
state, polling the CC lines causes the chip to be awaken from sleep
mode increasing the power usage. When partner is connected, or any
other CC lines change happens, the chip informs about it with alert.
It can be cached and used instead of asking the chip directly.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

sambhurst commented on subsys/usb/usb_c/usbc_tc_src_states.c:
```diff
@@ -295,6 +295,7 @@ void tc_attached_src_exit(void *obj)
 	tc_pd_enable(dev, false);
 
 	/* Stop sourcing VBUS */
+	__ASSERT(data->policy_cb_src_en != NULL, "Callback pointer should not be NULL");
```
This assertion is already done on line 291. Maybe you could do something like the following:
```
int policy_cb_src_en(dev, bool en)
{
    struct usbc_port_data *data = dev->data;

     __ASSERT(data->policy_cb_src_en != NULL, "policy_cb_src_en must not be NULL");

    return data->policy_cb_src_en(dev, en);
 }
````
And just use the new function for VBUS control.

semihalf-barnas-michal replied:
We talked internally and decided that this callback should be made optional, since it can be unused when the TCPC/PPC controls the power paths. So asserts will be removed and ifs with checks added instead.

sambhurst commented on include/zephyr/drivers/usb_c/usbc_tcpc.h:
```diff
@@ -460,12 +460,11 @@ static inline int tcpc_set_roles(const struct device *dev,
  * @param type  pointer to where message type is written. Can be NULL
  *
- * @retval true if message is pending, else false
+ * @retval 1 if there is pending message, else 0
```
I think we should stick to using the Zephyr return codes. What about returning -ENODATA for no pending message, and 0 for pending message?

semihalf-barnas-michal replied:
Sure, I've changed existing code and this new driver

sambhurst commented on subsys/usb/usb_c/usbc_tc_common.c:
```diff
@@ -215,14 +215,23 @@ static void tc_cc_open_entry(void *obj)
 
 	/* Disable VCONN */
-	tcpc_set_vconn(tcpc, false);
+	ret = tcpc_set_vconn(tcpc, false);
```
We should think about what to do if the tcpc_set_* functions fail. Maybe log the error as you're doing and then go to the error recovery state, but I'm not sure? Let's discuss offline.

semihalf-barnas-michal replied:
We should set the state to error recovery in both states

semihalf-barnas-michal replied:
As we discussed, the TCPC init failure will disable the type-c layer, but other failures during the runtime will move to the error recovery state.

sambhurst commented on subsys/usb/usb_c/usbc_tc_common.c:
```diff
@@ -141,6 +141,8 @@ static int tc_init(const struct device *dev)
 		data->policy_cb_src_en(dev, false);
 	}
 
+	tcpc_set_src_ctrl(tcpc, false);
```
Add comment

semihalf-barnas-michal commented on subsys/usb/usb_c/usbc_tc_src_states.c:
```diff
@@ -288,14 +288,13 @@ void tc_attached_src_exit(void *obj)
 
 	/* Stop sourcing VBUS */
-	data->policy_cb_src_en(dev, false);
+	if (data->policy_cb_src_en) {
```
Should the wrapper return 0 in case of no callback, or -ENOSYS, so the if would check for "0 or -ENOSYS"?

semihalf-barnas-michal commented on subsys/usb/usb_c/usbc_prl.c:
```diff
@@ -1175,9 +1175,11 @@ static void prl_rx_wait_for_phy_message(const struct device *dev)
 
 	/* Get the message */
-	if (tcpc_receive_data(tcpc, rx_emsg) <= 0) {
+	pending_rx = tcpc_is_rx_pending_msg(tcpc, NULL);
```
We should make some driver functions required, to not return -ENOSYS but add ASSERT to make sure they are implemented and then this suggestion can be applied.

semihalf-barnas-michal commented on subsys/usb/usb_c/usbc_tc_common.c:
```diff
@@ -55,7 +55,11 @@ void tc_run(const struct device *dev, const int32_t dpm_request)
 	case SM_INIT:
 		/* Initialize the Type-C layer */
-		tc_init(dev);
+		if (tc_init(dev) < 0) {
```
Yes, we should go to the disabled state here

barnas-michal committed:
usbc: merge the is_rx and get_rx_pending_msg functions

These two functions are used together so there is no need for
splitting them into two functions. This commit also makes this
function required to be implemented by the TCPC driver.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add helper function to enable the source path

Due to different possible ways of enabling the source path for
the USB-C port, this commit adds generic helper function that
tries to use callback and/or TCPC API function. In case of error, it
returns any of the error, except for the -ENOSYS from the TCPC.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add checks for results of TCPC functions

This commit adds checks for results and in case of error, handles it.
The failed initialization of the TCPC can postpone initialization in
case the TCPC needs more time after powering up. Otherwise if other
error happened, the Type-C Layer will be disabled.
Errors in other places of the USB-C stack, like communication errors,
moves the state machine to the error recovery, restarting the
communication with parner.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
usbc: add support for TCPC control of vbus sinking

Adds calls to the set_snk_ctrl of TCPC driver to enable and disable
the VBUS sinking. If the TCPC doesn't support these functions,
no errors should be printed.

Signed-off-by: Michał Barnaś <mb@semihalf.com>
